### PR TITLE
Card payments improvements

### LINKED ISSUE #27: Card payments improvements
As we're currently in the rapid development stage, we delivered a happy path for card payments. This task is aiming to implement missing features and provide fixes proposed during the code review.

## Small wins 🏆
- [x] All relevant fixes suggested in #22 are implemented
- [x] The card data form is validated
  - [x] A cardholder name cannot be empty, and should have at least 3 characters
  - [x] A card number must be exactly 16 characters long
  - [x] A CVV code must be exactly 3 characters long
  - [x] A month is selected from a dropdown to prevent mistakes
  - [x] A year is selected from a dropdown to prevent mistakes (a current year and next 15 years)

> *Important!* The validation must be done using some JS code or library, as we do not (and we cannot) pass the card data to the backend. No card data is touching the backend.

### PR DESCRIPTION
**Pull request by jakubtobiasz**
Closes #27

### CODE REVIEW

File: translations/messages.pl.yaml [hunk 1]
```diff
@@ -34,6 +34,9 @@ commerce_weavers_sylius_tpay:
                         october: 'Październik'
                         november: 'Listopad'
                         december: 'Grudzień'
+                    month_placeholder: 'Wybierz miesiąc'
+                    year: Rok
```
Comment (arti0090) [inline translations/messages.pl.yaml:38]:
```suggestion
                    year: 'Rok'
```
And for messages.en.yaml it could be changed as well :)

**Review by arti0090 — APPROVED** [review]

File: src/Form/Type/TpayCardType.php [hunk 2]
```diff
@@ -72,15 +73,29 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
...
 
         $builder->addModelTransformer($this->cardTypeDataTransformer);
     }
+
+    private function getCardValidYearsRange(): array
+    {
+        $result = [];
+        $currentYear = (int) date('Y');
+
+        foreach (range($currentYear, $currentYear + 10) as $year) {
```
Comment (coldic3) [inline src/Form/Type/TpayCardType.php:95]: Why `+ 10`? Is it only our assumption, or the max value is documented somewhere? Nevertheless, I'd move `10` the const.

File: src/Tpay/PayGroup.php [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+declare(strict_types=1);
+
+namespace CommerceWeavers\SyliusTpayPlugin\Tpay;
```
Comment (coldic3) [inline src/Tpay/PayGroup.php:5]:
```suggestion
namespace CommerceWeavers\SyliusTpayPlugin\Tpay\Enum;
```

WDYT?

File: src/Tpay/PayGroup.php [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+class PayGroup
```
Comment (coldic3) [inline src/Tpay/PayGroup.php:7]:
```suggestion
final class PayGroup
```

WDYT? I can't imagine a situation where extending these classes is needed.

File: translations/validators.pl.yaml [hunk 5]
```diff
@@ -7,3 +7,9 @@ commerce_weavers_sylius_tpay:
             encoded_card_data:
                 length: 'Dane karty nie mogą być puste.'
                 required: 'Dane karty są wymagane.'
+            card:
+                cvc: 'Kod CVC musi składać się z 3 cyfr.'
+                expiration_month: 'Miesiąc ważności musi być późniejszy lub równy bieżącemu miesiącowi.'
+                expiration_year: 'Rok ważności musi być późniejszy lub równy bieżącemu roku.'
```
Comment (coldic3) [inline translations/validators.pl.yaml:13]:
```suggestion
                expiration_year: 'Rok ważności musi być bieżący lub późniejszy.'
```

WDYT? Similar with `expiration_month`.

File: translations/validators.en.yaml [hunk 6]
```diff
@@ -7,3 +7,9 @@ commerce_weavers_sylius_tpay:
             encoded_card_data:
                 length: 'The card data cannot be empty.'
                 required: 'The card data is required.'
+            card:
+                cvc: 'The CVC must be composed of 3 digits.'
+                expiration_month: 'The expiry month must be later than or equal to the current month.'
+                expiration_year: 'The expiry year must be later than or equal to the current year.'
```
Comment (coldic3) [inline translations/validators.en.yaml:13]:
```suggestion
                expiration_year: 'The expiration year must be the current year or later.'
```
or if we follow more British English 
```suggestion
                expiration_year: 'The expiry year must be the current year or later.'
```

File: assets/shop/js/card_form.js [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+
+      event.target.value = value.slice(0, 3);
+    });
+  }
+
+  #registerCardNumberFormatting() {
+    this.#cardNumber.addEventListener('keypress', (event) => {
+      const value = event.target.value.replace(/\s/g, '');
+
+      if (value.length >= 16) {
```
Comment (coldic3) [inline assets/shop/js/card_form.js:42]: I'm for keeping or integers (except some simple one like 0, 1) in consts. Then you can reuse it in lines 54, 60.

So for example:
```suggestion
      if (value.length >= CARD_NUMBER_LENGHT) {
```

File: assets/shop/js/card_form.js [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+    this.#cvc = this.#form.querySelector('[data-tpay-cvc]');
+    this.#encryptedCard = this.#form.querySelector('[data-tpay-encrypted-card]');
+    this.#expirationMonth = this.#form.querySelector('[data-tpay-expiration-month]');
+    this.#expirationYear = this.#form.querySelector('[data-tpay-expiration-year]');
+    this.#submitButton = this.#form.querySelector('[type="submit"]');
+    this.#registerEvents();
+    this.#registerFormatters();
+  }
+
+  #registerCvcFormatting() {
```
Comment (coldic3) [inline assets/shop/js/card_form.js:30]: I'd put all private methods after public ones.

File: assets/shop/js/card_form.js [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+
+      if (newCursorPosition < 0) {
+        newCursorPosition = 0;
+      }
+
+      event.target.setSelectionRange(newCursorPosition, newCursorPosition);
+    });
+  }
+
+  isCardHolderNameValid() {
```
Comment (coldic3) [inline assets/shop/js/card_form.js:69]:
```suggestion
  #isCardHolderNameValid() {
```

Can it be private?

**Review by coldic3 — APPROVED** [review]

File: assets/shop/js/card_form.js [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+      }
+    });
+
+    this.#cardNumber.addEventListener('keyup', (event) => {
+      const currentPositionStart = event.target.selectionStart;
+      const currentLength = event.target.value.length;
+
+      const value = event.target.value.replace(/\s/g, '');
+
+      const parts = value.slice(0, 16).match(/\d{1,4}/g) || [];
```
Comment (coldic3) [inline assets/shop/js/card_form.js:56]:
```suggestion
      const parts = value.slice(0, MAX_CARD_NUMBER_LENGTH).match(/\d{1,4}/g) || [];
```

let's reuse the const

File: assets/shop/js/card_form.js [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+  }
+
+  isCvcValid() {
+    const regex = new RegExp(/^\d{3}$/);
+
+    return regex.test(this.getCardCvc());
+  }
+
+  isCardNumberValid() {
+    const regex = new RegExp(/^\d{16}$/);
```
Comment (coldic3) [inline assets/shop/js/card_form.js:82]:
```suggestion
    const regex = new RegExp(`/^\d{${MAX_CARD_NUMBER_LENGTH}}$/`);
```

**Review by coldic3 — APPROVED** [review]

**Pull request merged**